libcaliptra/examples/generic/main.c

```diff
@@ -5,22 +5,98 @@
 #include <string.h>
 #include <errno.h>
 #include <unistd.h>
+#include <openssl/evp.h>
 
 #include "caliptra_api.h"
+#include "caliptra_image.h"
 
 // Interface defined values
 extern struct caliptra_fuses  fuses;        // Device-specific location of Caliptra fuse data
 extern struct caliptra_buffer image_bundle; // Device-specific location of Caliptra firmware
 
-int main(int argc, char *argv[])
+static const uint32_t default_uds_seed[] = { 0x00010203, 0x04050607, 0x08090a0b, 0x0c0d0e0f,
+                                             0x10111213, 0x14151617, 0x18191a1b, 0x1c1d1e1f,
+                                             0x20212223, 0x24252627, 0x28292a2b, 0x2c2d2e2f };
+
+static const uint32_t default_field_entropy[] = { 0x80818283, 0x84858687, 0x88898a8b, 0x8c8d8e8f,
+                                                  0x90919293, 0x94959697, 0x98999a9b, 0x9c9d9e9f };
+
+static int set_fuses(struct caliptra_image_manifest *image)
 {
     int status;
 
+    const EVP_MD *md;
+    EVP_MD_CTX *ctx;
+
+    uint32_t md_vendor_pubkey[SHA384_DIGEST_WORD_SIZE];
+    uint32_t md_owner_pubkey[SHA384_DIGEST_WORD_SIZE];
+    int      mdlen;
+
+    OpenSSL_add_all_digests();
+
+    md = EVP_get_digestbyname("SHA384");
+
+    if (!md)
+    {
+        printf("Failed to acquire SHA384 digest\n");
+        return -1;
+    }
+
+    ctx = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(ctx, md, NULL);
+    EVP_DigestUpdate(ctx, &image->preamble.vendor_pub_keys, sizeof(struct image_vendor_pubkeys));
+    EVP_DigestFinal_ex(ctx, (unsigned char*)md_vendor_pubkey, &mdlen);
+    EVP_MD_CTX_destroy(ctx);
+
+    if (mdlen != SHA384_DIGEST_BYTE_SIZE)
+    {
+        printf("SHA384 digest from OpenSSL is not the correct size! e: %u a: %u", SHA384_DIGEST_BYTE_SIZE, mdlen);
+    }
+
+    ctx = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(ctx, md, NULL);
```

Comment:
openssl provides a utility function for all this:

https://github.com/openssl/openssl/blob/39ed7636e0d8a90512e7ccb811cd0bfcb7a79650/include/openssl/sha.h#L132

Reply:
When I went looking for that, their own documentation said "Applications should use the higher level functions [EVP_DigestInit(3)](https://www.openssl.org/docs/man1.0.2/man3/EVP_DigestInit.html) etc. instead of calling the hash functions directly."

https://www.openssl.org/docs/man1.0.2/man3/SHA384.html

I can swap over to calling them directly if that's preferred.

Reply:
No, I didn't realize that SHA384() doesn't just call the same implementation as EVP_DigestInit, sigh... Leave it as is.